### PR TITLE
Load submission for tasks

### PR DESCRIPTION
**Pull request by victorbalan**

### CODE REVIEW

File: controllers/submission.go [hunk 1]
```diff
@@ -122,14 +125,141 @@ func PostSubmission(ctx context.Context, w http.ResponseWriter, r *http.Request)
...
+		GetAll(ctx, &submissions)
+	if err != nil {
+		return http.StatusInternalServerError, err
+	}
+
+	json.NewEncoder(w).Encode(submissions.Key(keys))
+	return http.StatusOK, nil
+}
+
+type SubmissionWithUrl struct {
```
Comment (flowlo) [inline controllers/submission.go]: Don't export this.

File: util/storage.go [hunk 2]
```diff
@@ -57,6 +57,27 @@ func Load(ctx context.Context, bucket, name string) (io.ReadCloser, error) {
 	return fromStorage(ctx, bucket, name)
 }
 
+// TODO rename
+func LoadMultiURL(ctx context.Context, bucket, name string) ([]string, error) {
```
Comment (flowlo) [inline util/storage.go]: `ExposePrefix`?

File: test/robot.go [hunk 3]
```diff
@@ -164,13 +171,14 @@ func testRobot(m Map, in string) (moves []TypedPosition, err error) {
...
+		case "right":
 			direction = turnRight(direction)
-		case "LEFT":
+		case "left":
 			direction = turnLeft(direction)
-		case "PICK":
+		case "pick":
 			moves = pick(moves, m.Coins)
 		default:
+			fmt.Println("Bad command ", c)
```
Comment (flowlo) [inline test/robot.go]: Should be removed or be a `log.Debugf`

File: test/robot.go [hunk 4]
```diff
@@ -22,20 +22,25 @@ func init() {
...
-	if stdin, err = util.Load(util.CloudContext(ctx), sub.Code.Bucket, path.Dir(sub.Code.Name)+"/"+util.FileNames["robot"]); err != nil {
-		return
-	}
-	defer stdin.Close()
-
 	var testMapBytes, stdinBytes []byte
-	if stdinBytes, err = ioutil.ReadAll(stdin); err != nil {
-		return
+	tr := tar.NewReader(ball)
+	for {
```
Comment (flowlo) [inline test/robot.go]: That loop looks a bit crooked. You expect only one file, so you could do `tr.Next()` expect no `io.EOF`, then process stuff, then do `tr.Next()` again and expect `io.EOF`. Anyway, if there are multiple files and no error you'll write to `stdinBytes` multiple times which could be confusing to debug.

File: model/junit_test_result.go [hunk 5]
```diff
@@ -15,7 +15,8 @@ type JunitTestResult struct {
 	Start time.Time `datastore:",index"`
 	End   time.Time `datastore:",index"`
 
-	Results UnitTestResults `datastore:",noindex"`
+	Results  UnitTestResults `datastore:",noindex"`
+	Endpoint string
```
Comment (flowlo) [inline model/junit_test_result.go]: What is this endpoint used for again? To render results afterwards? Is it used currently?

File: controllers/submission.go [hunk 6]
```diff
@@ -123,14 +126,139 @@ func PostSubmission(ctx context.Context, w http.ResponseWriter, r *http.Request)
...
+	if !ok {
+		return http.StatusUnauthorized, nil
+	}
+
+	submissionKey, err := datastore.DecodeKey(mux.Vars(r)["key"])
+	if err != nil {
+		return http.StatusNotFound, err
+	}
+
+	keys, err := datastore.NewQuery("").Ancestor(submissionKey).Filter("__key__ >", submissionKey).KeysOnly().GetAll(ctx, nil)
```
Comment (flowlo) [inline controllers/submission.go]: Split to multiple lines please ...

File: controllers/submission.go [hunk 7]
```diff
@@ -123,14 +126,139 @@ func PostSubmission(ctx context.Context, w http.ResponseWriter, r *http.Request)
...
+		var results model.JunitTestResults
+		_, err = datastore.NewQuery(keys[0].Kind()).Ancestor(submissionKey).GetAll(ctx, &results)
+		if err != nil {
+			return http.StatusInternalServerError, err
+		}
+		json.NewEncoder(w).Encode(results)
+		return http.StatusOK, nil
+	case model.DiffTestResultKind:
+		var results model.DiffTestResults
+		_, err = datastore.NewQuery(keys[0].Kind()).Ancestor(submissionKey).GetAll(ctx, &results)
```
Comment (flowlo) [inline controllers/submission.go]: Split to multiple lines please ...

File: controllers/submission.go [hunk 8]
```diff
@@ -123,14 +126,139 @@ func PostSubmission(ctx context.Context, w http.ResponseWriter, r *http.Request)
...
+	}
+	if len(keys) == 0 {
+		json.NewEncoder(w).Encode([]string{})
+		return http.StatusOK, nil
+	}
+
+	switch keys[0].Kind() {
+	case model.JunitTestResultKind:
+		var results model.JunitTestResults
+		_, err = datastore.NewQuery(keys[0].Kind()).Ancestor(submissionKey).GetAll(ctx, &results)
```
Comment (flowlo) [inline controllers/submission.go]: Split to multiple lines please ...

File: controllers/submission.go [hunk 9]
```diff
@@ -123,14 +126,139 @@ func PostSubmission(ctx context.Context, w http.ResponseWriter, r *http.Request)
...
+	_, ok := passenger.FromContext(ctx)
+	if !ok {
+		return http.StatusUnauthorized, nil
+	}
+
+	resultKey, err := datastore.DecodeKey(mux.Vars(r)["resultKey"])
+	if err != nil {
+		return http.StatusNotFound, err
+	}
+	taskKey, err := datastore.DecodeKey(mux.Vars(r)["taskKey"])
```
Comment (flowlo) [inline controllers/submission.go]: Empty line after error handling block please.

File: model/junit_test_result.go (hunk 5 above)
Comment (victorbalan) [inline model/junit_test_result.go]: yep. It is used

File: controllers/submission.go [hunk 10]
```diff
@@ -123,14 +126,148 @@ func PostSubmission(ctx context.Context, w http.ResponseWriter, r *http.Request)
...
+			Ancestor(submissionKey).
+			GetAll(ctx, &results)
+		if err != nil {
+			return http.StatusInternalServerError, err
+		}
+		json.NewEncoder(w).Encode(results)
+		return http.StatusOK, nil
+	}
+
+	json.NewEncoder(w).Encode([]string{})
```
Comment (flowlo) [inline controllers/submission.go]: `w.Write([]byte("[]"))` would be a lot cheaper?!

File: controllers/submission.go [hunk 11]
```diff
@@ -123,14 +126,148 @@ func PostSubmission(ctx context.Context, w http.ResponseWriter, r *http.Request)
...
+		GetAll(ctx, nil)
+	if err != nil {
+		return http.StatusInternalServerError, err
+	}
+	if len(keys) == 0 {
+		json.NewEncoder(w).Encode([]string{})
+		return http.StatusOK, nil
+	}
+
+	switch keys[0].Kind() {
```
Comment (flowlo) [inline controllers/submission.go]: You could write the empty array in a `default` case.

File: test/robot.go [hunk 12]
```diff
@@ -22,20 +22,21 @@ func init() {
...
 	if testMap, err = util.Load(util.CloudContext(ctx), util.TemplateBucket, t.Params["tests"]); err != nil {
 		return
 	}
 	defer testMap.Close()
 
-	if stdin, err = util.Load(util.CloudContext(ctx), sub.Code.Bucket, path.Dir(sub.Code.Name)+"/"+util.FileNames["robot"]); err != nil {
-		return
+	var testMapBytes, stdinBytes []byte
+
+	tr := tar.NewReader(ball)
```
Comment (flowlo) [inline test/robot.go]: `tr` is never closed. Same for `ball`.

File: controllers/submission.go [hunk 13]
```diff
@@ -123,14 +126,148 @@ func PostSubmission(ctx context.Context, w http.ResponseWriter, r *http.Request)
...
+		_, err = datastore.NewQuery(keys[0].Kind()).
+			Ancestor(submissionKey).
+			GetAll(ctx, &results)
+		if err != nil {
+			return http.StatusInternalServerError, err
+		}
+		json.NewEncoder(w).Encode(results)
+		return http.StatusOK, nil
+	default:
+		json.NewEncoder(w).Encode([]byte("[]"))
```
Comment (flowlo) [inline controllers/submission.go]: `[]` is valid JSON AFAIK, why encode it again? I bet you'd get `["[]"]`

File: controllers/submission.go [hunk 14]
```diff
@@ -123,14 +126,148 @@ func PostSubmission(ctx context.Context, w http.ResponseWriter, r *http.Request)
...
+		return http.StatusOK, nil
```
Comment (flowlo) [inline controllers/submission.go]: You `return http.StatusOK, nil` in *all* cases, so move it after the `switch`.

File: controllers/submission.go [hunk 15]
```diff
@@ -123,14 +126,148 @@ func PostSubmission(ctx context.Context, w http.ResponseWriter, r *http.Request)
...
+			Ancestor(submissionKey).
+			GetAll(ctx, &results)
+		if err != nil {
+			return http.StatusInternalServerError, err
+		}
+		json.NewEncoder(w).Encode(results)
+		return http.StatusOK, nil
+	default:
+		w.Write([]byte("[]"))
+		return http.StatusOK, nil
```
Comment (flowlo) [inline controllers/submission.go]: You return `return http.StatusOK, nil` in all cases, so move it after the switch.

**Pull request merged**